webodf/lib/ops/OperationTransformMatrix.js

```diff
@@ -274,6 +274,262 @@ ops.OperationTransformMatrix = function OperationTransformMatrix() {
     /* Transformation methods */
 
     /**
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpecA
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpecB
+     * @param {!boolean} hasAPriority
+     * @return {?{opSpecsA:!Array.<!Object>, opSpecsB:!Array.<!Object>}}
+     */
+    function transformAddAnnotationAddAnnotation(addAnnotationSpecA, addAnnotationSpecB, hasAPriority) {
+        var aBeforeB;
+
+        if (addAnnotationSpecA.position < addAnnotationSpecB.position) {
+            aBeforeB = true;
+        } else if (addAnnotationSpecB.position < addAnnotationSpecA.position) {
+            aBeforeB = false;
+        } else {
+            aBeforeB = hasAPriority;
+        }
+
+        if (aBeforeB) {
+            if (addAnnotationSpecB.position < addAnnotationSpecA.position + addAnnotationSpecA.length) {
+                addAnnotationSpecA.length += 2;
+            }
+            addAnnotationSpecB.position += 2;
+        } else {
+            if (addAnnotationSpecA.position < addAnnotationSpecB.position + addAnnotationSpecB.length) {
+                addAnnotationSpecB.length += 2;
+            }
+            addAnnotationSpecA.position += 2;
+        }
+
+        return {
+            opSpecsA:  [addAnnotationSpecA],
+            opSpecsB:  [addAnnotationSpecB]
+        };
+    }
+
+    /**
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpec
+     * @param {!ops.OpApplyDirectStyling.Spec} applyDirectStylingSpec
+     * @return {?{opSpecsA:!Array.<!Object>, opSpecsB:!Array.<!Object>}}
+     */
+    function transformAddAnnotationApplyDirectStyling(addAnnotationSpec, applyDirectStylingSpec) {
+        if (addAnnotationSpec.position <= applyDirectStylingSpec.position) {
+            applyDirectStylingSpec.position += 2;
+        } else if (addAnnotationSpec.position <= applyDirectStylingSpec.position + applyDirectStylingSpec.length) {
+            applyDirectStylingSpec.length += 2;
+        }
+
+        return {
+            opSpecsA:  [addAnnotationSpec],
+            opSpecsB:  [applyDirectStylingSpec]
+        };
+    }
+
+    /**
+     * @param {!ops.OpAddAnnotation.Spec} addAnnotationSpec
+     * @param {!ops.OpInsertText.Spec} insertTextSpec
+     * @return {?{opSpecsA:!Array.<!Object>, opSpecsB:!Array.<!Object>}}
+     */
+    function transformAddAnnotationInsertText(addAnnotationSpec, insertTextSpec) {
+        if (insertTextSpec.position <= addAnnotationSpec.position) {
```

Comment:
Part of me wants to wrap this up in a standard function, as it's an often repeated pattern here, and having it common would allow some easier high-level checks.

Possibly ``moveOrExpandRegions``? Would make reviewing easier too!